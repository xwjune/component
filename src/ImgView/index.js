/**
 * 移动端图片全屏预览
 *
 * @author 小巷 <xwjune@163.com>
 * @date 2019/01/17
 */
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import ImgCatch from '../ImgCatch';
import styles from './index.scss';

export default class ImgView extends React.PureComponent {
  static propTypes = {
    src: PropTypes.string,
    onError: PropTypes.func,
  };

  state = {
    isErr: false,
  };

  componentWillUnmount() {
    if (this.view) {
      this.handleHide();
    }
  }

  onError = () => {
    this.setState({
      isErr: true,
    });
    if (this.props.onError) {
      this.props.onError();
    }
  };

  renderView = (src) => {
    return (
      <div
        className={styles.view}
        onClick={this.handleHide}
      >
        <img src={src} alt="预览图片" />
      </div>
    );
  };

  handleShow = (e) => {
    e.stopPropagation();
    e.preventDefault();
    if (this.props.src && !this.state.isErr) {
      this.view = document.createElement('div');
      document.body.appendChild(this.view);
      let style = document.body.getAttribute('style');
      if (style) {
        style += `${/;$/.test(style.trimEnd()) ? '' : ';'} overflow: hidden;`;
      } else {
        style = 'overflow: hidden;';
      }
      document.body.setAttribute('style', style);
      ReactDOM.render(this.renderView(this.props.src), this.view);
    }
  };

  handleHide = () => {
    let style = document.body.getAttribute('style');
    style = style.replace(/\s?overflow: hidden;/, '');
    document.body.setAttribute('style', style);
    document.body.removeChild(this.view);
    this.view = null;
  };

  render() {
    return (
      <ImgCatch
        {...this.props}
        onError={this.onError}
        onClick={this.handleShow}
      />
    );
  }
}
