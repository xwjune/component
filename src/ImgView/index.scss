.view {
  position: fixed;
  z-index: 99999;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: #000;
  img {
    max-width: 100%;
    max-height: 100%;
  }
}
