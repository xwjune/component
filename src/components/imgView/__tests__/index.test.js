import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import sinon from 'sinon';
import ImgView from '../index';

describe('ImgView', () => {
  beforeAll(() => {
    document.body.removeAttribute('style');
  });

  test('renders correctly', () => {
    const wrapper = shallow(
      <ImgView
        src="pic.png"
        width="100"
        height="100"
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('全屏预览', () => {
    const wrapper = shallow(
      <ImgView src="pic.png" />
    );
    // 点击放大
    wrapper.find('img').simulate('click');
    expect(wrapper.instance().view).not.toBeUndefined();
    expect(document.body.getAttribute('style')).toBe('overflow: hidden;');
    expect(document.body.querySelectorAll('.jun-img-view')).toHaveLength(1);

    // 全屏快照
    expect(wrapper.instance().view).toMatchSnapshot();
    expect(document.body).toMatchSnapshot();

    // 点击还原
    document.body.querySelector('.jun-img-view').click();
    expect(wrapper.instance().view).toBeUndefined();
    expect(document.body.getAttribute('style')).toBe('');
    expect(document.body.querySelectorAll('.jun-img-view')).toHaveLength(0);
  });

  test('body样式', () => {
    document.body.setAttribute('style', 'background: #fff');
    const wrapper = mount(<ImgView src="pic.png" />);
    wrapper.find('img').simulate('click');
    expect(document.body.getAttribute('style')).toBe('background: #fff; overflow: hidden;');
    wrapper.unmount();
  });

  test('body样式2', () => {
    document.body.setAttribute('style', 'background: #fff;');
    const wrapper = mount(<ImgView src="pic.png" />);
    wrapper.find('img').simulate('click');
    expect(document.body.getAttribute('style')).toBe('background: #fff; overflow: hidden;');
    wrapper.unmount();
  });

  test('图片异常点击不放大', () => {
    const wrapper = shallow(<ImgView src="pic.png" />);
    wrapper.find('img').simulate('error');
    wrapper.find('img').simulate('click');
    expect(wrapper.instance().view).toBeUndefined();
    expect(document.body.querySelectorAll('.jun-img-view')).toHaveLength(0);
  });

  test('componentWillUnmount', () => {
    const wrapper = shallow(<ImgView src="pic.png" />);
    sinon.spy(ImgView.prototype, 'componentWillUnmount');
    wrapper.find('img').simulate('click');
    expect(ImgView.prototype.componentWillUnmount.calledOnce).toBeFalsy();
    wrapper.unmount();
    expect(ImgView.prototype.componentWillUnmount.calledOnce).toBeTruthy();
    expect(document.body.querySelectorAll('.jun-img-view')).toHaveLength(0);
  });
});
