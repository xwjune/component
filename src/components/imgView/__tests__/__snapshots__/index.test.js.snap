// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgView renders correctly 1`] = `
<img
  alt="图片"
  height="100"
  src="pic.png"
  width="100"
/>
`;

exports[`ImgView 全屏预览 1`] = `
<div>
  <div
    class="jun-img-view"
  >
    <img
      alt="预览图片"
      src="pic.png"
    />
  </div>
</div>
`;

exports[`ImgView 全屏预览 2`] = `
<body
  style="overflow: hidden;"
>
  <div>
    <div
      class="jun-img-view"
    >
      <img
        alt="预览图片"
        src="pic.png"
      />
    </div>
  </div>
</body>
`;
